### PR TITLE
SDK-1.0.0-alpha

### PR DESCRIPTION
**Pull request by lukeg90**

### CODE REVIEW

File: src/channels/CredentialChannels/backgroundCredentialChannel.ts [hunk 1]
```diff
@@ -1,5 +1,16 @@
+import type {
+  IEncryptedMessage,
+  IReject,
+} from '@kiltprotocol/kilt-extension-api/types';
```
Comment (arty-name) [inline src/channels/CredentialChannels/backgroundCredentialChannel.ts:4]: `@kiltprotocol/kilt-extension-api` is a funny name

**Comment by lukeg90** [comment]
Still need to fix the tests

File: src/views/SignDid/SignDid.tsx [hunk 2]
```diff
@@ -52,11 +51,9 @@ export function SignDid({ identity, popupData, onCancel, credentials }: Props) {
       const { didDocument, authenticationKey } =
         await getIdentityCryptoFromSeed(seed);
 
-      const signature = Utils.Crypto.u8aToHex(
-        authenticationKey.sign(plaintext),
-      );
+      const signature = Crypto.u8aToHex(authenticationKey.sign(plaintext));
       const didKeyUri =
-        `${didDocument.uri}${didDocument.authentication[0].id}` as DidResourceUri;
+        `${didDocument.id}${didDocument.authentication?.[0]}` as DidUrl;
```
Comment (arty-name) [inline src/views/SignDid/SignDid.tsx:56]: The authentication key is optional now?

File: src/utilities/did/did.ts [hunk 3]
```diff
@@ -1,54 +1,80 @@
...
 
   return {
     ...parsed,
     fullDid,
   };
 }
+
+export function computeKeyId(key: Uint8Array): VerificationMethod['id'] {
+  return `#${blake2AsHex(key, 256)}`;
+}
```
Comment (arty-name) [inline src/utilities/did/did.ts:70]: If the SDK people are asking us to attach this gun to our foot, we should really push back. How come application code has to know KILT internals in such depth?

File: src/utilities/getDeposit/getDeposit.ts [hunk 4]
```diff
@@ -17,61 +25,59 @@ async function getDefaultDeposit() {
...
+  did: Did | undefined,
 ): Promise<DepositData | undefined> {
   const api = ConfigService.get('api');
 
-  if (!did || Did.parse(did).type === 'light') {
+  if (!did || parse(did).type === 'light') {
     if ('deposit' in api.consts.did) {
       // TODO: remove this `if` once the Spiritnet is updated
-      return { amount: api.consts.did.deposit as BN };
+      return { amount: api.consts.did.deposit as unknown as BN };
```
Comment (arty-name) [inline src/utilities/getDeposit/getDeposit.ts:66]: It could be a good moment to do what is to be done.

File: src/utilities/getDeposit/getDeposit.ts [hunk 5]
```diff
@@ -17,61 +25,59 @@ async function getDefaultDeposit() {
...
     }
 
-    const baseDeposit = api.consts.did.baseDeposit as BN;
-    const keyDeposit = api.consts.did.keyDeposit as BN;
+    const baseDeposit = api.consts.did.baseDeposit as unknown as BN;
+    const keyDeposit = api.consts.did.keyDeposit as unknown as BN;
```
Comment (arty-name) [inline src/utilities/getDeposit/getDeposit.ts:70]: I believe something is wrong with the SDK types if we have to do such type casts

File: src/utilities/getTabEncryption/getTabEncryption.ts [hunk 6]
```diff
@@ -45,34 +52,43 @@ export async function getTabEncryption(
...
-    `${sporranDidDocument.uri}${sporranEncryptionDidKey.id}` as DidResourceUri;
+    `${sporranDidDocument.id}${sporranEncryptionDidKey}` as DidUrl;
+
+  const { contentStream } = await DidResolver.dereference(
+    dAppEncryptionKeyUri,
+    {},
+  );
+  const dAppEncryptionDidKey = contentStream as VerificationMethod | undefined;
+
+  if (dAppEncryptionDidKey?.type !== 'Multikey') {
```
Comment (arty-name) [inline src/utilities/getTabEncryption/getTabEncryption.ts:77]: This new code hints that a convenience method is needed in the SDK

File: src/utilities/getTabEncryption/getTabEncryption.ts [hunk 7]
```diff
@@ -45,34 +52,43 @@ export async function getTabEncryption(
...
+
   await verifyDidConfigResource(dAppDid, sender.url);
 
   async function decrypt(encrypted: IEncryptedMessage): Promise<IMessage> {
     return Message.decrypt(
       encrypted,
-      async ({ data: box, peerPublicKey, nonce }) => {
-        const data = Utils.Crypto.decryptAsymmetric(
+      async ({ data: box, peerPublicKey, nonce }: DecryptRequestData) => {
+        const data = Crypto.decryptAsymmetric(
```
Comment (arty-name) [inline src/utilities/getTabEncryption/getTabEncryption.ts:91]: Why isn’t the type of the callback argument inferred automatically?

File: src/utilities/getTabEncryption/getTabEncryption.ts [hunk 8]
```diff
@@ -86,27 +102,32 @@ export async function getTabEncryption(
...
+    );
+
+    return {
+      data: box,
+      nonce,
+      keyUri: sporranEncryptionDidKeyUri,
+    };
+  }
+
+  // TODO: use encryption from kilt-extension-api library
```
Comment (arty-name) [inline src/utilities/getTabEncryption/getTabEncryption.ts:119]: I’ll leave this comment as a reminder about the TODO

File: src/utilities/identities/identities.ts [hunk 9]
```diff
@@ -308,14 +303,16 @@ export async function importIdentity(
   const seed = mnemonicToMiniSecret(backupPhrase);
 
   const lightDidDocument = getLightDidFromSeed(seed);
-  const lightDid = lightDidDocument.uri;
+  const lightDid = lightDidDocument.id;
+
+  const {
+    didDocumentMetadata: { canonicalId, deactivated },
+  } = await DidResolver.resolve(lightDid, {});
```
Comment (arty-name) [inline src/utilities/identities/identities.ts:310]: Isn’t the second parameter optional? I feel that it has to be.

File: src/views/SignDidExtrinsic/SignDidExtrinsic.tsx [hunk 10]
```diff
@@ -250,45 +257,46 @@ export function SignDidExtrinsic({ identity }: Props) {
...
         throw new Error('No key relationship found');
       }
 
       if (keyRelationship === 'assertionMethod') {
         const api = ConfigService.get('api');
-        const attestationKey = deriveAttestationKeyFromSeed(seed);
+        const keypair = deriveAttestationKeyFromSeed(seed);
+        const attestationKey = verificationMethodFromKeypair(keypair, did);
+
+        fullDidDocument.verificationMethod?.push(attestationKey);
```
Comment (arty-name) [inline src/views/SignDidExtrinsic/SignDidExtrinsic.tsx:271]: Let’s put a comment here about why we’re certain this array exists

File: src/utilities/didUpgrade/didUpgrade.ts [hunk 11]
```diff
@@ -25,33 +33,51 @@ import { makeFakeIdentityCrypto } from '../makeFakeIdentityCrypto/makeFakeIdenti
...
-  sign: SignExtrinsicCallback,
+  signers: SignerInterface[],
   submitter?: KiltAddress,
 ): Promise<DidTransaction> {
-  const { fullDid: did } = parseDidUri(lightDidDocument.uri);
+  const { fullDid: did } = parseDidUri(lightDidDocument.id);
 
-  const extrinsic = await Did.getStoreTx(
-    lightDidDocument,
+  await DidResolver.dereference(lightDidDocument.id, {});
```
Comment (arty-name) [inline src/utilities/didUpgrade/didUpgrade.ts:51]: The intention of calling this is unclear, should it be explained in a comment?

**Review by arty-name — APPROVED** [review]

File: src/utilities/identities/identities.ts (hunk 9 above)
Comment (lukeg90) [reply src/utilities/identities/identities.ts:310]: I also think it's weird that it's not optional, especially when it's called `ResolutionOptions`

Comment (arty-name) [reply src/utilities/identities/identities.ts:310]: Let’s apply slight pressure to someone

File: src/views/SignDid/SignDid.tsx (hunk 2 above)
Comment (lukeg90) [reply src/views/SignDid/SignDid.tsx:56]: Apparently

File: src/utilities/did/did.ts (hunk 3 above)
Comment (lukeg90) [reply src/utilities/did/did.ts:70]: In that case the key ID has to be computed beforehand so it can be passed as a signer to the batch transaction.

File: src/utilities/identities/identities.ts (hunk 9 above)
Comment (lukeg90) [reply src/utilities/identities/identities.ts:310]: @rflechtner we are applying slight pressure to you 🙂

File: src/views/SignDid/SignDid.tsx (hunk 2 above)
Comment (lukeg90) [reply src/views/SignDid/SignDid.tsx:56]: Maybe @rflechtner can give us some insight

Comment (lukeg90) [reply src/views/SignDid/SignDid.tsx:56]: I expect the reason is that the DID specification says authentication keys are optional

File: src/utilities/did/did.ts (hunk 3 above)
Comment (arty-name) [reply src/utilities/did/did.ts:70]: That’s an unfortunate outcome of the design

File: src/utilities/didUpgrade/didUpgrade.ts (hunk 11 above)
Comment (rflechtner) [reply src/utilities/didUpgrade/didUpgrade.ts:51]: I agree, can't see how this would do anything? If you wanted to check for errors, you'd have to look at the dereferenceMetadata.error property

File: src/views/SignDid/SignDid.tsx (hunk 2 above)
Comment (rflechtner) [reply src/views/SignDid/SignDid.tsx:56]: Yeah resolving a deactivated DID still returns a DID document, but an empty one (only the id property is there).

File: src/utilities/didUpgrade/didUpgrade.ts (hunk 11 above)
Comment (lukeg90) [reply src/utilities/didUpgrade/didUpgrade.ts:51]: Not sure where this line came from, obviously doesn't belong here 😅

File: src/views/SignDid/SignDid.tsx (hunk 2 above)
Comment (arty-name) [reply src/views/SignDid/SignDid.tsx:56]: Feels like purity has beat practicality in this case

File: src/utilities/did/did.ts (hunk 3 above)
Comment (rflechtner) [reply src/utilities/did/did.ts:70]: This shouldn't be necessary. The result of you adding an assertionMethod in the same transaction in which you're using it is that you'll have to pass a 'doctored' DID document to the helper that selects the right signer with which to authorise  the call - the function relies on the mapping between purposes and verification methods in the DID document. Because you're messing with the document you can use whatever id you choose, as long as it corresponds to the id on the signer for the key pair you're using here. So may as well be `#assert` or something

File: src/utilities/getDeposit/getDeposit.ts (hunk 5 above)
Comment (rflechtner) [reply src/utilities/getDeposit/getDeposit.ts:70]: yeah why do you have to force this?

File: src/utilities/identities/identities.ts (hunk 9 above)
Comment (rflechtner) [reply src/utilities/identities/identities.ts:310]: That is an interesting one! I agree this is massively annoying and I'm inclined to make it optional - especially bc it is only required on the typing, but the actual implementation handles omitting it just fine. A quick look at the typings explains what feels like an oversight (Antonio built this, so I went to a similar discovery process):

```
  resolve: (
    /**
     * This is the DID to resolve.
     * This input is REQUIRED and the value MUST be a conformant DID as defined in 3.1 DID Syntax.
     */
    did: Did,
    /**
     * A metadata structure containing properties defined in 7.1.1 DID Resolution Options.
     * This input is REQUIRED, but the structure MAY be empty.
     */
    resolutionOptions: ResolutionOptions
  ) => Promise<ResolutionResult>
```
Meaning: it is a specification thing, which is why I am unsure what to do about it. I don't see any harm in making it optional really, but for some reason it's explicitly required even if empty...

Comment (arty-name) [reply src/utilities/identities/identities.ts:310]: Okay, my opinion about this specification authors has sunk a little more, nothing to see here

Comment (rflechtner) [reply src/utilities/identities/identities.ts:310]: I think we don't need to be sticklers about it really, I'll put it in my list of improvements

File: src/utilities/getDeposit/getDeposit.ts (hunk 5 above)
Comment (lukeg90) [reply src/utilities/getDeposit/getDeposit.ts:70]: Oh this is a remnant of when the api types weren't working for me

File: src/utilities/getTabEncryption/getTabEncryption.ts (hunk 6 above)
Comment (lukeg90) [reply src/utilities/getTabEncryption/getTabEncryption.ts:77]: I'm not super confident that this is actually needed here. @rflechtner what do you think?

File: src/channels/CreateDidChannels/types.ts [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
+
 import { HexString } from '@polkadot/util/types';
 
 import { DAppName } from '../../dApps/AccessChannels/DAppName';
 import { Origin } from '../../dApps/AccessChannels/Origin';
 
 export type CreateDidInput = DAppName & {
   submitter: KiltAddress;
-  pendingDidUri?: DidUri;
+  pendingDidUri?: Did;
```
Comment (ntn-x2) [inline src/channels/CreateDidChannels/types.ts:10]: Would changing `pendingDidUrl` to `pendingDid` following the new naming convention of the SDK break anything?

File: src/channels/CredentialChannels/backgroundCredentialChannel.ts (hunk 1 above)
Comment (ntn-x2) [reply src/channels/CredentialChannels/backgroundCredentialChannel.ts:4]: Agree!

File: src/channels/SignDidChannels/types.ts [hunk 13]
```diff
@@ -1,18 +1,19 @@
...
 import { HexString } from '@polkadot/util/types';
-import { DidResourceUri, DidUri, ICredential } from '@kiltprotocol/sdk-js';
 
 import { DAppName } from '../../dApps/AccessChannels/DAppName';
 import { Origin } from '../../dApps/AccessChannels/Origin';
 
 export type SignDidInput = DAppName & {
   plaintext: string;
-  didUri?: DidUri;
+  didUri?: Did;
```
Comment (ntn-x2) [inline src/channels/SignDidChannels/types.ts:10]: Same here, and pretty much anywhere else.

File: src/channels/SignDidChannels/types.ts [hunk 14]
```diff
@@ -1,18 +1,19 @@
...
 };
 
 export type SignDidOriginInput = SignDidInput & Origin;
 
 export interface SignDidOutput {
   signature: HexString;
-  didKeyUri: DidResourceUri;
+  didKeyUri: DidUrl;
```
Comment (ntn-x2) [inline src/channels/SignDidChannels/types.ts:17]: Just a heads-up, mind that the new definition of `DidUrl` can include optional query parameters, so something like `did:kilt:4....?param=value#keyId` would now be a valid URL. They can of course simply be ignored, but I wanted to raise awareness that that's also possible now.

File: src/utilities/did/did.ts [hunk 15]
```diff
@@ -1,54 +1,80 @@
...
 }
 
-export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
-  const api = ConfigService.get('api');
-  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
-    .document;
+export async function getFullDidDocument(did: Did): Promise<DidDocument> {
+  const { didDocument } = await DidResolver.resolve(did, {});
+  if (!didDocument) {
+    throw new Error('Unable to resolve DID');
```
Comment (ntn-x2) [inline src/utilities/did/did.ts:32]: There's even resolution metadata which gives you the exact reason why the DID could not be resolved. Maybe it would be useful to add it to the error thrown.

File: src/utilities/did/useIsOnChainDidDeleted.ts [hunk 16]
```diff
@@ -1,18 +1,19 @@
...
+import { DidResolver } from '@kiltprotocol/sdk-js';
 
 import { useAsyncValue } from '../useAsyncValue/useAsyncValue';
 
-async function getIsOnChainDidDeleted(
-  did: DidUri | undefined,
-): Promise<boolean> {
+async function getIsOnChainDidDeleted(did: Did | undefined): Promise<boolean> {
   if (!did) {
     return false;
```
Comment (ntn-x2) [inline src/utilities/did/useIsOnChainDidDeleted.ts:9]: I know it was here before, but basically if the input DID is nullish, it is considered **not** deleted? As I read the logic below, this function returns `true` if the DID is valid, so the naming of the function is misleading since it gives away the opposite, as far as I can see. Or otherwise, the conditions below are a hard to follow for someone like me 😄 Maybe the boolean condition can be split up into two.

File: src/utilities/getDeposit/getDeposit.ts (hunk 4 above)
Comment (ntn-x2) [reply src/utilities/getDeposit/getDeposit.ts:66]: Confirm. We don't have this deposit logic anymore. It can go.

File: src/utilities/getTabEncryption/getTabEncryption.ts (hunk 6 above)
Comment (ntn-x2) [reply src/utilities/getTabEncryption/getTabEncryption.ts:77]: @arty-name what would the convenience method do?

File: src/utilities/identities/identities.ts (hunk 9 above)
Comment (ntn-x2) [reply src/utilities/identities/identities.ts:310]: Not an oversight, but compliance to the resolution spec. We could make it optional, I agree.

File: src/injectedScript.ts [hunk 17]
```diff
@@ -52,7 +53,7 @@ async function storeMessageFromSporran(
 
 async function startSession(
   unsafeDAppName: string,
-  dAppEncryptionKeyId: DidResourceUri,
+  dAppEncryptionKeyId: DidUrl,
```
Comment (ntn-x2) [inline src/injectedScript.ts:56]: The naming `keyId` reflects the old naming scheme. At the very least, I would change it to `keyUri`.

File: src/injectedScript.ts [hunk 18]
```diff
@@ -148,19 +149,17 @@ async function startSession(
...
-> {
+function getDidList(): ReturnType<InjectedSporranApi<unknown>['getDidList']> {
   const dAppName = document.title.substring(0, 50);
   return injectedShareIdentitiesChannel.get({ dAppName });
 }
 
 async function signWithDid(
   plaintext: string,
-  didUri?: DidUri,
+  didUri?: Did,
```
Comment (ntn-x2) [inline src/injectedScript.ts:159]: This also should simply be `did`

File: src/injectedScript.ts [hunk 19]
```diff
@@ -185,7 +184,7 @@ async function signExtrinsicWithDid(
 
 async function getSignedDidCreationExtrinsic(
   submitter: KiltAddress,
-  pendingDidUri?: DidUri,
+  pendingDidUri?: Did,
```
Comment (ntn-x2) [inline src/injectedScript.ts:187]: Also here.

Comment (lukeg90) [reply src/injectedScript.ts:187]: I'm not concerned with changing any of the naming in this PR. It's just to test how to alpha works with Sporran and will not be merged into the main branch

File: src/injectedScript.ts (hunk 17 above)
Comment (lukeg90) [reply src/injectedScript.ts:56]: I'm not concerned with changing any of the naming since the purpose of this branch is just to see how the SDK alpha works with Sporran.

There are also a number of places where changing the naming would break other apps that interact with the Sporran API.

Comment (ntn-x2) [reply src/injectedScript.ts:56]: Ok so these comments will be gone? I think it makes sense to leave the ones that would break otherwise, and change the ones that don't break.

Comment (lukeg90) [reply src/injectedScript.ts:56]: I'm not merging this branch

Comment (ntn-x2) [reply src/injectedScript.ts:56]: Ok, it wasn't clear. I was asked to review, and I reviewed. Feel free to ignore most of the comments then.

Comment (lukeg90) [reply src/injectedScript.ts:56]: And I don't agree that having naming inconsistencies makes sense

File: src/utilities/getTabEncryption/getTabEncryption.ts (hunk 6 above)
Comment (arty-name) [reply src/utilities/getTabEncryption/getTabEncryption.ts:77]: It would get a key of desired type or throw. Having to do type casting and checking the key type as above is not something to be written every time one gets a key, in my opinion :)

File: src/utilities/did/useIsOnChainDidDeleted.ts (hunk 16 above)
Comment (lukeg90) [reply src/utilities/did/useIsOnChainDidDeleted.ts:9]: I agree it's confusing, but I tried not to change any of Sporran's logic unless the new SDK made it necessary to change it.

File: src/utilities/getTabEncryption/getTabEncryption.ts (hunk 6 above)
Comment (rflechtner) [reply src/utilities/getTabEncryption/getTabEncryption.ts:77]: a verification method to public key function maybe? Would be great if that was able to turn both the old and the new verification method type into a combination of public key bytes and a key type identifier, such as ed25519 / sr25519 / secp256k1 / x25519 etc.

File: src/views/SignDidExtrinsic/SignDidExtrinsic.tsx (hunk 10 above)
Comment (lukeg90) [reply src/views/SignDidExtrinsic/SignDidExtrinsic.tsx:271]: Now I'm wondering if I'm certain it exists 😅 It must exist because all DIDs have at least 1 verification method unless they're deactivated, right?

Comment (ntn-x2) [reply src/views/SignDidExtrinsic/SignDidExtrinsic.tsx]: Yes, and that would be the authentication key.

Comment (arty-name) [reply src/views/SignDidExtrinsic/SignDidExtrinsic.tsx]: I also thought that, but scanning through code should not cause such mental exercises, so a comment should say this :)
